File: src/csharp/binpl/binpl.cs
/*

BINPL (v1.0-preview1 / July 12, 2024)
Copyright (c) 2024 Cyril John Magayaga

*/
using System;
using System.IO;

public class Binpl
{
    private const string BINPL_VERSION = "v1.0-preview1";

    public static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: binpl [binary file] [-d or --decimal] [-h or --hexadecimal]");
            return;
        }

        if (args.Length == 1 && (args[0] == "-h" || args[0] == "--help"))
        {
            PrintUsage();
            return;
        }

        string filePath = args[0];

        if (!File.Exists(filePath))
        {
            Console.WriteLine($"Error: Unable to open file '{filePath}'");
            return;
        }

        string binaryCode = File.ReadAllText(filePath);

        int decimalMode = 0; // Default mode: ASCII interpretation
        int hexadecimalMode = 0; // Default mode: ASCII interpretation

        if (args.Length > 1)
        {
            for (int i = 1; i < args.Length; i++)
            {
                if (args[i] == "-d" || args[i] == "--decimal")
                {
                    decimalMode = 1;
                }
                
                else if (args[i] == "-h" || args[i] == "--hexadecimal")
                {
                    hexadecimalMode = 1;
                }
                
                else if (args[i] == "-v" || args[i] == "--version")
                {
                    Console.WriteLine(BINPL_VERSION);
                    return;
                }
                
                else if (args[i] == "--author")
                {
                    Console.WriteLine("Copyright (c) 2024 Cyril John Magayaga");
                    return;
                }
                
                else
                {
                    Console.WriteLine($"Error: Invalid option '{args[i]}'");
                    PrintUsage();
                    return;
                }
            }
        }

        // Parse binary code, skipping comments
        string[] lines = binaryCode.Split(new[] { "\r\n", "\r", "\n" }, StringSplitOptions.None);
        foreach (string line in lines)
        {
            if (!line.StartsWith(";;"))
            {
                InterpretBinary(line, decimalMode, hexadecimalMode);
            }
        }
    }

    private static void InterpretBinary(string binaryCode, int decimalMode, int hexadecimalMode)
    {
        int len = binaryCode.Length;

        // Make sure the length is a multiple of 8
        if (len % 8 != 0)
        {
            Console.WriteLine("Error: Invalid binary string length");
            Environment.Exit(1);
        }

        // Iterate through the binary string, interpreting each group of 8 bits
        for (int i = 0; i < len; i += 8)
        {
            int decimalValue = 0;
            for (int j = 0; j < 8; j++)
            {
                decimalValue = (decimalValue << 1) | (binaryCode[i + j] - '0');
            }

            if (decimalMode != 0)
            {
                Console.Write($"{decimalValue} ");
            }
            
            else if (hexadecimalMode != 0)
            {
                Console.Write($"{decimalValue:X2} ");
            }
            
            else
            {
                Console.Write($"{(char)decimalValue}");
            }
        }
        Console.WriteLine();
    }

    private static void PrintUsage()
    {
        Console.WriteLine("Usage: binpl [binary file] [-d or --decimal] [-h or --hexadecimal]");
        Console.WriteLine("       -v or --version       Version information.");
        Console.WriteLine("       -h or --help          Usage information.");
        Console.WriteLine("       -d or --decimal       Interpret binary as decimal numbers.");
        Console.WriteLine("       -h or --hexadecimal   Interpret binary as hexadecimal numbers.");
        Console.WriteLine("       --author              Author information.");
    }
}

File: src/ts/deno/binpl.ts
/*

BINPL (v1.0-preview1 / July 12, 2024)
Copyright (c) 2024 Cyril John Magayaga

*/
const BINPL_VERSION = "v1.0-preview1";

// Function to interpret binary code
function interpretBinary(binaryCode: string, decimalMode: boolean, hexadecimalMode: boolean): void {
    const len = binaryCode.length;

    // Make sure the length is a multiple of 8
    if (len % 8 !== 0) {
        console.error("Error: Invalid binary string length");
        Deno.exit(1);
    }

    // Iterate through the binary string, interpreting each group of 8 bits
    for (let i = 0; i < len; i += 8) {
        let decimalValue = 0;
        for (let j = 0; j < 8; j++) {
            decimalValue = (decimalValue << 1) | (binaryCode.charCodeAt(i + j) - "0".charCodeAt(0));
        }
        if (decimalMode) {
            Deno.stdout.write(new TextEncoder().encode(decimalValue + " "));
        }
        
        else if (hexadecimalMode) {
            Deno.stdout.write(new TextEncoder().encode(decimalValue.toString(16).toUpperCase() + " "));
        }
        
        else {
            Deno.stdout.write(new TextEncoder().encode(String.fromCharCode(decimalValue)));
        }
    }
    console.log();
}

// Main function
async function main(): Promise<number> {
    const args = Deno.args;
    
    if (args.length !== 1 && args.length !== 2) {
        console.log("Usage: binpl [binary file]");
        return 1;
    }

    const option = args[0];
    
    if (option === "-h" || option === "--help") {
        console.log("Usage: binpl [binary file] [-d or --decimalMode] [-h or --hexadecimal]");
        console.log("       -v or --version       Version information.");
        console.log("       -h or --help          Usage information.");
        console.log("       -d or --decimal       Interpret binary as decimal numbers.");
        console.log("       -h or --hexadecimal   Interpret binary as hexadecimal numbers.");
        console.log("       --author              Author information.");
        return 0;
    } 
    
    if (option === "-v" || option === "--version") {
        console.log(BINPL_VERSION);
        return 0;
    }
    
    if (option === "--author") {
        console.log("Copyright (c) 2024 Cyril John Magayaga");
        return 0;
    }

    const filePath = args[0];
    let binaryCode: string;
    try {
        binaryCode = await Deno.readTextFile(filePath);
    }
    
    catch (error) {
        console.error(`Error: Unable to open file "${filePath}"`);
        return 1;
    }

    let decimalMode = false; // Default mode: ASCII interpretation
    let hexadecimalMode = false; // Default mode: ASCII interpretation

    if (args.length === 2) {
        if (args[1] === "-d" || args[1] === "--decimal") {
            decimalMode = true;
        }
        
        else if (args[1] === "-h" || args[1] === "--hexadecimal") {
            hexadecimalMode = true;
        }
        
        else {
            console.error(`Error: Invalid option "${args[1]}"`);
            return 1;
        }
    }

    // Parse binary code, skipping comments
    const lines = binaryCode.split("\n");
    for (const line of lines) {
        if (!line.startsWith(";;")) {
            interpretBinary(line, decimalMode, hexadecimalMode);
        }
    }

    return 0;
}

// Run main function
main();
